### PR TITLE
Add job status endpoint to server

### PR DESCRIPTION
**Pull request by val500**
## Description

This adds an endpoint to Testflinger server that receives status updates from Testflinger agents. These updates require a webhook that the server will forward the rest of the update message to.

## Documentation

Documentation was added to job_schema.rst to reflect new job_status_webhook option in the job yaml. The server ReadMe was updated with details on the new status endpoint. 

## Web service API changes
- Interface: [POST] /v1/status
- Request Structure: 
    ```json
    {
        "agent_id": "<string>",
        "job_queue": "<string>",
        "job_status_webhook": "<URL as string>",
        "events": [
            {
  	        "event_name": "<string enum of events>",
  	        "timestamp": "<string>",
  	        "detail_msg": "<string>"
            },
            ...
         ]
    }

    ```

- Response Status Codes:
    - HTTP 200 (OK)
    - HTTP 400 (Bad request) - The arguments could not be processed by the server
    - HTTP 504 (Gateway Timeout) - The webhook URL timed out
- Rationale: This endpoint receives status updates from the agent if there is a webhook specified in the job yaml and posts the update to this webhook.

## Tests

For the server, unit tests were added to test_v1.py. For the agents, unit tests were added to both test_client.py and test_agent.py to test both webhook posting behavior and overall status updating.

### CODE REVIEW

File: server/src/api/v1.py
Comment (omar-selo) [inline server/src/api/v1.py:512]: Would be great if the structure of the request and response are [schemas](https://github.com/canonical/testflinger/blob/main/server/src/api/schemas.py) rather than a comment. Schemas are exposed on testflinger's [api docs](https://testflinger.canonical.com/docs). And [apiflask](https://apiflask.com/) would provide automatic request validation and serialization

File: server/src/api/v1.py
Comment (omar-selo) [inline server/src/api/v1.py:521]: why does it timeout after 1 second? it seems more reliable if it has retries or even better adds it to a queue to send it to TO reliably

**Review by omar-selo — CHANGES_REQUESTED** [review]
Main concern is reliability, as in I wonder if we should add this to a queue rather than forward immediately

File: server/src/api/v1.py
Comment (val500) [reply server/src/api/v1.py:521]: I think we timeout in order to prevent the server from hanging for too long on any individual request. We include previous phase results in each status update so if any of them fail, you can still get the information in the next one. I think @plars can expand here.

As for the queue, we had discussed using something like celery to maintain a queue of these updates, but I think @mz2 had raised some concerns about adding extra dependencies.

Comment (mz2) [reply server/src/api/v1.py]: > I think we timeout in order to prevent the server from hanging for too long on any individual request. 

Sorry, I'm not very clear on this detail. Omar's on point about the good options: either retry, or dequeue and mark the queue event done when it was successfully delivered. In absence of a queue, retry seems saner than dropping with a 1s timeout. The server doesn't have to and therefore shouldn't try to worry about the request / response cycle or conserving itself for concurrency (add more workers if needed).

If we have elsewhere code that optimises for closing a request / response cycle quickly in case of a failure that is retryable, we should similarly not go there anywhere else either.

Let's talk at the sprint about this.

**Review by plars — CHANGES_REQUESTED** [review]
Looks like you're missing the enums.py at the very least. Can you add that here or tell me where to find it if it's external?

File: agent/testflinger_agent/stop_condition_checkers.py
Comment (omar-selo) [inline agent/testflinger_agent/stop_condition_checkers.py:1]: This is a tough one. Reading the stop condition checkers makes me feel that changing the whole agent flow to be event driven sounds better. The idea is if something happens like a job is cancelled, then different objects that have different responsibilities can react to this event (without mixing responsibilities together). The runner would cleanup the process. The logger would print some logs. The event emitter would forward this event to the server. But this would be a big refactor that we need to think more carefully about before doing. So I don't recommend trying to do that now. Instead, perhaps it's enough to keep the checkers as they were before. Then at the runner code, if a stop condition returns a string then we emit an appropriate event depending on the type of the checker. e.g. it he type is `JobCancelledChecker` then we `emitJobCancelled` (using the same event emitter class). What I'm not sure about though is whether `emitJobCancelled` should be triggered from the runner or the agent.

**Review by omar-selo — CHANGES_REQUESTED** [review]

File: server/src/api/v1.py
Comment (omar-selo) [inline server/src/api/v1.py:521]: doesn't sound right to have a 1s timeout, I'd like to push again towards a queue

File: common/README.rst
Comment (plars) [reply common/README.rst:1]: Yeah, I think the idea was to land that PR and then rebase this on main? Did you cherry-pick it here or something @val500 ?

File: agent/testflinger_agent/job.py
Comment (val500) [reply agent/testflinger_agent/job.py:129]: exit_event and exit_reason are the event_name and detail_msg fields respectively. They get used in agent.py when the job gets run.

File: server/src/api/v1.py
Comment (omar-selo) [reply server/src/api/v1.py:521]: Alternatively, if we can do something similar to [background tasks](https://fastapi.tiangolo.com/reference/background/) in fastapi

File: agent/testflinger_agent/job.py
Comment (plars) [reply agent/testflinger_agent/job.py:129]: They were landed as part of a previous PR to make it easier to find this information, so that it could be returned when the phase finishes running. This way we know the reason it stopped and can send that up to the server to be logged, stored, or forwarded to test-observer as needed from one place

File: server/src/api/schemas.py
Comment (plars) [reply server/src/api/schemas.py:180]: it's already in the job data that it received when starting the job. By transmitting it at the time that we tell the server to send this, we save the server from having to do another lookup to find it from the job data.

File: agent/testflinger_agent/client.py
Comment (plars) [reply agent/testflinger_agent/client.py:383]: I think this is actually ok. While it's true that we DO want to expose some information about recent provision failures on the testflinger UI under the agent detail page, I think it will have to work a little different. Instead of discrete events like this, we are really only concerned with why the provision phase finished, and probably storing it in a way that is easier to retrieve when we retrieve the other agent data. I'm going to be looking into that piece soon, and if it turns out that we can reuse this then I'll certainly try to do so. But the call for sending stuff along to test-observer is already a little strange and probably best to keep it separate for now.

File: common/README.rst
Comment (val500) [reply common/README.rst:1]: Rebased on main now

File: agent/testflinger_agent/client.py
Comment (omar-selo) [reply agent/testflinger_agent/client.py:383]: It does feel strange though that the agent would know about the webhook. wdyt?

File: server/src/api/v1.py
Comment (val500) [reply server/src/api/v1.py:521]: I think for now we'll do a 3 second timeout with 3 retries. With regards to queues, there was some direction to use Temporal over Celery, but setting up that system might be better to do in a separate issue.

File: agent/testflinger_agent/agent.py
Comment (nadzyah) [inline agent/testflinger_agent/agent.py:275]: Why is it an empty string? We can probably set the default value to be “” instead of requiring to pass it

Comment (val500) [reply agent/testflinger_agent/agent.py]: Changed to default to ""

File: agent/testflinger_agent/client.py
Comment (val500) [reply agent/testflinger_agent/client.py]: Changed to 30 seconds

File: server/src/api/v1.py
Comment (omar-selo) [reply server/src/api/v1.py]: sure, as long as we do go back to it. "for now" solutions tend to become "forever" :sweat_smile:

**Review by omar-selo — DISMISSED** [review]
I did have a few comments left, but to avoid blocking anything I'll approve. As I think this is a good first step

**Review by plars — CHANGES_REQUESTED** [review]
I have a few more comments below. I think it would be nice to add a bit more on the unit tests, especially on the possible error/failure conditions. And I would definitely recommend and end-to-end test of this now that you have the test-observer side prototyped.

File: agent/testflinger_agent/runner.py
Comment (val500) [reply agent/testflinger_agent/runner.py:92]: That's a good catch - I think it worked because I check if the stop_event is None in the run function.

**Comment by omar-selo** [comment]
It occurred to me that it would be useful for TO if TF informs it about the last event. As in TF can mark an event as `last` to inform listeners that there will be no more events emitted after it. That way TO can say that this test execution has ended. Otherwise TO will need to know in detail which events would be final events. I prefer if TO doesn't actually know what the events mean for now. wdyt @plars @val500 ?

**Comment by val500** [comment]
> It occurred to me that it would be useful for TO if TF informs it about the last event. As in TF can mark an event as `last` to inform listeners that there will be no more events emitted after it. That way TO can say that this test execution has ended. Otherwise TO will need to know in detail which events would be final events. I prefer if TO doesn't actually know what the events mean for now. wdyt @plars @val500 ?
> 
> Update: Just had a meeting with Andrej and Matias about these events. Here are some notes:
> 
>     * Would be great to have traceability, such that we can get to a job from TO through the events. This could be in the form of a url.
> 
>     * TO needs more information about the events, namely it needs to know if an event is the last event, and if an event is a failure/cancellation. We don't want to deduce these on TO by the event name because that couples TO to TF. Instead it would be useful if along with every event perhaps a couple of booleans are set, something like `isFinal` and `isFailure`.
> 
> 
> wdyt @plars @val500

I think every job will end with the cleanup phase regardless of whether the job was cancelled or something threw an error, so the last event will always be ended_cleanup - @plars can confirm

**Comment by plars** [comment]
> @omar-selo I think every job will end with the cleanup phase regardless of whether the job was cancelled or something threw an error, so the last event will always be ended_cleanup - @plars can confirm

That's true for testflinger specifically, but the idea is to make it more generic on both sides. We could set something like a boolean "IsFinal" but for us, it would always be on that cleanup event because that's the last thing that would happen, regardless of whether any part of the job failed.

However, that means that something like an event showing that the job timed out, cancelled, etc... would have already happened. If this is what you are really interested in, then perhaps we should mark those with an "IsTerminal" event for event types that will wind up terminating the whole job? would that be enough for you to pick those out as the reason for the overall job termination?

File: agent/testflinger_agent/agent.py
Comment (plars) [inline agent/testflinger_agent/agent.py:208]: This is created too early and could leak into the next job. Delay creation of this event_emitter until we actually start processing a job.

File: agent/testflinger_agent/client.py
Comment (plars) [inline agent/testflinger_agent/client.py:369]: It's nice to add type annotations when we add new code at least. Can we go ahead and do that here like you did for EventEmitter?

**Review by plars — CHANGES_REQUESTED** [review]

File: agent/testflinger_agent/runner.py
Comment (plars) [inline agent/testflinger_agent/runner.py:90]: If we're going to do it this way, then I think we shouldn't just overload output to make it a tuple. Instead, maybe we could get `event, detail` = checker()` then you could specifically check `if event is not None:`...

File: agent/testflinger_agent/runner.py
Comment (plars) [inline agent/testflinger_agent/runner.py:91]: You still need to return 2 values here, not just 1. When we are calling self.check_stop_condition() on line 142, it's expecting both the event and the stop_reason (which should maybe be "detail" or something since now the event also conveys the reason for it)

**Review by plars — CHANGES_REQUESTED** [review]
Thanks for the updates! we should definitely go ahead and implement the things agreed to on the card unless you have further input on it.  A few more comments, but looking better :)

File: agent/testflinger_agent/agent.py
Comment (val500) [reply agent/testflinger_agent/agent.py:211]: I thought I had a case where the Job construction threw an exception which caused the event_emitter to not exist in the catch/finally block, but looking at it again, I'm not sure if Job construction can throw an error.

File: agent/testflinger_agent/runner.py
Comment (val500) [reply agent/testflinger_agent/runner.py:91]: I think output here is two values based on the return of the stop condition checkers, but I can make that more explicit.

**Review by plars — CHANGES_REQUESTED** [review]
Meeting notes: need `resource_url` added - we discussed adding a JOB_START event at the beginning like the JOB_END event we have at the end. The detail section for that JOB_START event can contain the `resource_url` which would just be `https://testflinger.canonical.com/jobs/<JOB_ID>`. That isn't going to change throughout the job, so there's no need to send it with every single event.

File: server/src/api/v1.py
Comment (plars) [inline server/src/api/v1.py:496]: After our chat today I was looking at how you would add the resource_url and realized that you don't have the job_id here anywhere. Then I got to thinking more about the path for this API and I'm not sure if `/v1/agents/status` is the right place (also the README is wrong but I'll comment on that separately). I think putting it in this path would hint that it's a status update about "agents in general". In the case of this API though, it's that at all. Instead, we're sending events that have happened in the context of a **job**. Sure, it has a specific agent associated with it, and a queue, but it's all data about what happened during the job itself. Also, associating it with a job_id means we can send that information across quite easily.

What do you think about putting it instead, at: `/job/<job_id>/events` or something like that? I'm open to other ideas, but I feel like this is at least a bit more consistent with what we already have, and conveys the meaning a bit better. Also, it's a pretty simple change, and it gets you the job_id that you're also going to want to insert as the `resource_url`

File: server/README.rst
Comment (plars) [inline server/README.rst:345]: See my comment about the path for this API in v1.py, and make sure the README is consistent with what we have there.

**Review by plars — CHANGES_REQUESTED** [review]

File: server/src/api/v1.py
Comment (val500) [reply server/src/api/v1.py:496]: Does the server need the job_id? If we're going to put the resource_url in the JOB_START event, it sounds easier to have the agent create the url with the job_id than have the server parse the events and create the url for the JOB_START event.

Comment (plars) [reply server/src/api/v1.py]: @val500 the main thing I was looking for here was a small change to the endpoint which it looks like you took care of. However, I don't see where that resource_url is actually injected?

Comment (val500) [reply server/src/api/v1.py]: It's in the detail for the JOB_START event

Comment (val500) [reply server/src/api/v1.py]: I think the URL is actually incorrect here - would it be {server_name}/job/{job_id}?

Comment (plars) [reply server/src/api/v1.py]: right now, it looks like you have `f"{self.client.server}/job/{job.job_id}/events"` but I think this is wrong, I was under the impression that this should be a link to the url for the job details page, which would be something like `f"{self.client.server}/jobs/{job.job_id}"` (note the "s" in "jobs" also). Let's sync up about this tomorrow as I think Omar also had some related comments on the TO side.

Comment (omar-selo) [reply server/src/api/v1.py]: Yes we'd need the job link similar to https://testflinger.canonical.com/jobs/697b8930-3d88-48d0-8458-aafe2820e312 for example

**Review by plars — APPROVED** [review]
This looks like it should be good, and thanks for the video showing your testing of it against the TO changes on the other PR.

As discussed yesterday, please wait until the TO changes land so that we can land these PRs at the same time, and test them in staging together.

**Pull request merged**
